rust-symcrypt/src/hash.rs

```diff
@@ -381,13 +391,26 @@ pub struct Sha256State(Pin<Box<symcrypt_sys::SYMCRYPT_SHA256_STATE>>);
 
 impl Sha256State {
     pub fn new() -> Self {
-        let mut instance = Sha256State(Box::pin(symcrypt_sys::SYMCRYPT_SHA256_STATE::default()));
+        let mut instance = Sha256State(Box::pin(Sha256InnerState{inner: symcrypt_sys::SYMCRYPT_SHA256_STATE::default(), _pinned: PhantomPinned}));
         unsafe {
             // SAFETY: FFI calls
-            symcrypt_sys::SymCryptSha256Init(&mut *instance.0);
+            symcrypt_sys::SymCryptSha256Init(instance.get_inner_mut());
         }
         instance
     }
+
+    // Safe method to access the inner state mutably
+    pub(crate) fn get_inner_mut(&mut self) -> *mut symcrypt_sys::SYMCRYPT_SHA256_STATE {
+        unsafe { 
+            // SAFETY: getting a mutable raw pointer to the inner state
+            &mut self.0.as_mut().get_unchecked_mut().inner as *mut _
+        }
+    }
+    
+    // Safe method to access the inner state immutably
+    pub(crate) fn get_inner(&self) -> *const symcrypt_sys::SYMCRYPT_SHA256_STATE {
+        &self.0.as_ref().get_ref().inner as *const _
+    }
```

Comment:
Incidentally, this function does not make me as nervous because a const pointer or reference cannot be used to move a value. That also shows up in the fact that you did need to use any `unsafe` in its body.